### ISSUE TITLE
[Question]Pipeline batches inferencing

### ISSUE THREAD

**Issue by adizhol** [pytorch/pytorch#32648]
Hi,
Google published a paper where they "pipeline" the inference (and backprop) stage (each later at a time).
[https://arxiv.org/abs/2001.06232](https://arxiv.org/abs/2001.06232)
I thought that most frameworks already do this during the forward pass when inferencing, since you don't need to wait for the batch to finish calculating all the outputs before loading a new batch.
If that's not the case, is any work on this planned for pytorch?

Thanks!

cc @ezyang @SsnL @albanD @zou3519 @gqchen @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @aazzolini @xush6528 @osalpekar @vincentqb

**Comment by vincentqb** [comment]
Tagging distributed too given that this also relates to Hogwild.

**Comment by vincentqb** [comment]
As far as I understand, one can do this as long as the layers are on different devices/machines, right?

**Comment by rmcavoy** [comment]
It definitely challenging to do forward+backward out of order efficiently but it is straightforward to make a dataloader that loads the next batch while the previous batch is still calculating following e.g. https://github.com/pytorch/pytorch/issues/15849#issuecomment-573921048. That dataloader just continuously will load samples into memory until a maximum of twice the batch size is reached (it also doesn't restart the workers every epochs which provides its own speedup). This kind of parallel dataloading can provide a substantial speedup for tasks where the dataloading and calculation are a similar length.

**Comment by adizhol** [comment]
> As far as I understand, this algorithm requires layers to be on different devices/machines, right?

I think that the first stage in developing something like this for inference-only should support a single device. This pipeline will fail if the GPU's resources are used by concurrent computations.

**Comment by adizhol** [comment]
@rmcavoy The idea is to start the calculations of the next batch while the previous batch is still calculating

**Comment by mrshenli** [comment]
Hey @adizhol 

> The idea is to start the calculations of the next batch while the previous batch is still calculating

If you are looking for GPipe-like parallelism, yes, it is possible and it's quite simple. Checkout this [tutorial](https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html).

The paper you cited seems to do it differently though:

>  we overwrite network activations whenever new ones, i.e., from new frames, become available.
Such a more gradual accumulation of information from both passes breaks the precise correspondence between gradients and activations, leading to theoretically more noisy weight updates.

I haven't read the full paper, but looks like they could use last batch's gradients and this batch's activations to update parameters? Is that correct?

**Comment by adizhol** [comment]
@mrshenli I wasn't familiar with this tutorial, so thank you! I will definitely read it :)

> I haven't read the full paper, but looks like they could use last batch's gradients and this batch's activations to update parameters? Is that correct?

Yes, the seem to be using the gradients of only one of the batches, arguing that it's ok when using videos since the changes in the images are assumed to be small. (I guess they could have also stored these gradients and used a running average)
They also hypothesize that this acts as a regularization for the training.

**Comment by vincentqb** [comment]
Thanks for the discussion! Based on the above, it seems like the discussion would be better continued on the forum if needed. Please feel free to re-open if you would like to request a specific enhancement or feature based on this.
